scanner/updater/export.go

```diff
@@ -0,0 +1,107 @@
+package updater
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/libvuln/jsonblob"
+	"github.com/quay/claircore/libvuln/updates"
+	"github.com/quay/zlog"
+	"github.com/stackrox/stackrox/scanner/v4/updater/manual"
+	"golang.org/x/time/rate"
+
+	_ "github.com/quay/claircore/updater/defaults"
+)
+
+// Export is responsible for triggering the updaters to download Common Vulnerabilities and Exposures (CVEs) data
+// and then outputting the result as a zstd-compressed file with .ztd extension
+func Export(ctx context.Context, outputDir string) error {
+	var outOfTree []driver.Updater
+
+	// Append updater sets directly to the outOfTree
+	appendUpdaterSet := func(updaterSet driver.UpdaterSet, err error) {
+		if err != nil {
+			zlog.Error(ctx).Err(err).Send()
+			return
+		}
+		outOfTree = append(outOfTree, updaterSet.Updaters()...)
+	}
+
+	appendUpdaterSet(manual.UpdaterSet(ctx, nil))
+	err := os.MkdirAll(outputDir, 0700)
+	if err != nil {
+		return err
+	}
+
+	// create temp file
+	outputFile, err := os.Create(filepath.Join(outputDir, "output.json.ztd"))
+	if err != nil {
+		return err
+	}
+	defer outputFile.Close()
+
+	limiter := rate.NewLimiter(rate.Every(time.Second), 5)
+	httpClient := &http.Client{
+		Transport: &rateLimitedTransport{
+			limiter:   limiter,
+			transport: http.DefaultTransport,
+		},
+	}
+
+	zstdWriter, err := zstd.NewWriter(outputFile)
+	if err != nil {
+		return err
+	}
+	defer zstdWriter.Close()
+
```

Comment:
@crozzy  I was having the workflow interruption problem while I am trying to dump all the vuln data to one json blob before writing to a ztd file. So I decided to keep using two json blobs. I was guessing it's either memory issue or the api we are calling has rate limiting.  If it were a rate limiting problem, we would likely see failure messages in the log. Strangely, there's no explicit messages provided for the job interruption in this case. [](https://github.com/stackrox/stackrox/actions/runs/5717948673/job/15492759478)

Reply:
Looks like OOM in the runner, which would make sense as the vulnerabilities are just being buffered in memory with the jsonblob store until they're written to files. Some routes forward:

- Keep it like this
- Work on increasing memory in the runner (self-hosted runner?)
- Rework the jsonblob store to be less memory intensive (not sure how currently)

I kind of wonder about the json produced from this process, won't it be two distinct root json objects?

Reply:
that's possible. But is it going to be a problem while being imported in Claircore offline mode?

Reply:
I confirmed with @crozzy  that the output file will look like:
```
{"name":"bad vuln"}
{"name":"worse vuln"}
```
so it's appendable cc: @RTann

Reply:
@jvdm for central downloading the json blob, that's 110 MB so I'm guessing that won't cause too much memory consumption in central but I could be wrong. 
For uncompressing the json, that's all done by Claircore offline import. So that would be the in the scope of investigating performance and scalability of Claircore.